File: examples/java/cases/tasks/task_list/TaskListApiCase.java
package cases.tasks.task_list;

import core.BaseApiCase;
import core.shared.AppBaseContext.AppLogger;
import core.shared.AppStructuralContracts.AppCaseError;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

public class TaskListApiCase
    extends BaseApiCase<TaskListDomainCase.TaskListInput, TaskListDomainCase.TaskListOutput> {
  public interface TaskStore {
    List<Map<String, Object>> read() throws Exception;
    void write(List<Map<String, Object>> value) throws Exception;
    void reset() throws Exception;
  }

  private final TaskListDomainCase domainCase = new TaskListDomainCase();

  public TaskListApiCase(ApiContext ctx) {
    super(ctx);
  }

  @Override
  public ApiResponse<TaskListDomainCase.TaskListOutput> handler(
      TaskListDomainCase.TaskListInput input
  ) throws Exception {
    ApiResponse<TaskListDomainCase.TaskListOutput> result =
        execute(input == null ? new TaskListDomainCase.TaskListInput() : input);

    if (result.success) {
      result.statusCode = 200;
      return result;
    }

    result.statusCode = mapErrorCodeToStatus(result.error == null ? null : result.error.code());
    return result;
  }

  @Override
  public Object router() {
    return Map.of(
        "method", "GET",
        "path", "/tasks",
        "handler", (RouteHandler) request -> handler(new TaskListDomainCase.TaskListInput())
    );
  }

  @Override
  public void test() throws Exception {
    InMemoryTaskStore store = new InMemoryTaskStore();
    store.reset();
    store.write(List.of(
        Map.of(
            "id", "task_001",
            "title", "Older task",
            "status", "todo",
            "createdAt", "2026-03-18T12:00:00.000Z",
            "updatedAt", "2026-03-18T12:00:00.000Z"
        ),
        Map.of(
            "id", "task_002",
            "title", "Newer task",
            "status", "doing",
            "createdAt", "2026-03-18T12:30:00.000Z",
            "updatedAt", "2026-03-18T12:30:00.000Z"
        )
    ));

    TaskListApiCase apiCase = new TaskListApiCase(testContext(store));
    ApiResponse<TaskListDomainCase.TaskListOutput> result = apiCase.handler(
        new TaskListDomainCase.TaskListInput()
    );

    if (!result.success || result.data == null) {
      throw new IllegalStateException("test: handler should return a successful task list");
    }

    if (result.data.tasks().size() != 2) {
      throw new IllegalStateException("test: task list should return all persisted tasks");
    }

    if (!"task_002".equals(result.data.tasks().get(0).id())) {
      throw new IllegalStateException("test: task list should sort by createdAt descending");
    }

    store.write(List.of(
        Map.of(
            "id", "broken",
            "title", "Broken task",
            "status", "broken",
            "createdAt", "2026-03-18T12:00:00.000Z",
            "updatedAt", "2026-03-18T12:00:00.000Z"
        )
    ));

    ApiResponse<TaskListDomainCase.TaskListOutput> invalidResult = apiCase.handler(
        new TaskListDomainCase.TaskListInput()
    );

    if (invalidResult.success) {
      throw new IllegalStateException("test: invalid persisted records must return failure");
    }
  }

  @Override
  protected void _validate(TaskListDomainCase.TaskListInput input) {
    try {
      domainCase.validate(input);
    } catch (Exception error) {
      throw new AppCaseError("VALIDATION_FAILED", error.getMessage());
    }
  }

  @Override
  protected Object _repository() {
    return resolveTaskStore();
  }

  @Override
  protected TaskListDomainCase.TaskListOutput _service(
      TaskListDomainCase.TaskListInput input
  ) throws Exception {
    List<Map<String, Object>> persisted = resolveTaskStore().read();
    List<TaskListDomainCase.Task> tasks;

    try {
      tasks = TaskListDomainCase.fromPersistedCollection(
          persisted,
          "task_list.persisted_tasks"
      );
    } catch (IllegalArgumentException error) {
      throw new AppCaseError("INTERNAL", error.getMessage());
    }

    tasks.sort(Comparator.comparing(task -> Instant.parse(task.createdAt())));
    List<TaskListDomainCase.Task> sorted = new ArrayList<>(tasks);
    sorted.sort((left, right) -> right.createdAt().compareTo(left.createdAt()));

    TaskListDomainCase.TaskListOutput output = new TaskListDomainCase.TaskListOutput(sorted);
    domainCase.validateOutput(output);
    return output;
  }

  private TaskStore resolveTaskStore() {
    Object providersValue = ctx.extra.get("providers");
    if (!(providersValue instanceof Map<?, ?> providers)) {
      throw new AppCaseError("INTERNAL", "task_list requires providers in ctx.extra");
    }

    Object taskStoreValue = providers.get("taskStore");
    if (!(taskStoreValue instanceof TaskStore taskStore)) {
      throw new AppCaseError("INTERNAL", "task_list requires a configured taskStore provider");
    }

    return taskStore;
  }

  private static int mapErrorCodeToStatus(String code) {
    if ("VALIDATION_FAILED".equals(code)) {
      return 400;
    }
    return 500;
  }

  private static ApiContext testContext(TaskStore store) {
    ApiContext context = new ApiContext();
    context.correlationId = "task-list-api-test";
    context.logger = new NoopLogger();
    context.extra.put("providers", Map.of("taskStore", store));
    return context;
  }

  private static final class InMemoryTaskStore implements TaskStore {
    private final List<Map<String, Object>> values = new ArrayList<>();

    @Override
    public List<Map<String, Object>> read() {
      return new ArrayList<>(values);
    }

    @Override
    public void write(List<Map<String, Object>> value) {
      values.clear();
      values.addAll(value);
    }

    @Override
    public void reset() {
      values.clear();
    }
  }

  private static final class NoopLogger implements AppLogger {
    @Override
    public void debug(String message, Map<String, Object> meta) {}

    @Override
    public void info(String message, Map<String, Object> meta) {}

    @Override
    public void warn(String message, Map<String, Object> meta) {}

    @Override
    public void error(String message, Map<String, Object> meta) {}
  }
}

File: examples/kotlin/cases/tasks/task_list/task_list.api.case.kt
package app.protocol.examples.kotlin.cases.tasks.task_list

import app.protocol.examples.kotlin.core.api.ApiContext
import app.protocol.examples.kotlin.core.api.ApiResponse
import app.protocol.examples.kotlin.core.api.BaseApiCase
import app.protocol.examples.kotlin.core.api.RouteBinding
import app.protocol.examples.kotlin.core.shared.AppCaseError
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject

private val json = Json { ignoreUnknownKeys = false }

private data class TaskStore(
    val read: suspend () -> List<JsonObject>,
    val write: suspend (List<JsonObject>) -> Unit,
    val reset: (suspend () -> Unit)?,
)

private fun mapErrorCodeToStatus(code: String?): Int =
    when (code) {
        "VALIDATION_FAILED" -> 400
        else -> 500
    }

class TaskListApi(
    ctx: ApiContext,
) : BaseApiCase<TaskListInput, TaskListOutput>(ctx) {
    private val domainCase = TaskListDomain()

    override val usesService: Boolean = true

    override suspend fun handler(input: TaskListInput): ApiResponse<TaskListOutput> {
        val result = execute(input)
        return if (result.success) {
            result.copy(statusCode = 200)
        } else {
            result.copy(statusCode = mapErrorCodeToStatus(result.error?.code))
        }
    }

    override fun router(): RouteBinding =
        RouteBinding(
            method = "GET",
            path = "/tasks",
            handler = { handler(TaskListInput) },
        )

    override suspend fun test() {
        val store = resolveTaskStore()
        store.reset?.invoke()
        store.write(
            listOf(
                json.encodeToJsonElement(
                    Task.serializer(),
                    Task(
                        id = "task_001",
                        title = "Older task",
                        status = "todo",
                        createdAt = "2026-03-18T12:00:00Z",
                        updatedAt = "2026-03-18T12:00:00Z",
                    ),
                ).jsonObject,
                json.encodeToJsonElement(
                    Task.serializer(),
                    Task(
                        id = "task_002",
                        title = "Newer task",
                        status = "doing",
                        createdAt = "2026-03-18T12:30:00Z",
                        updatedAt = "2026-03-18T12:30:00Z",
                    ),
                ).jsonObject,
            ),
        )

        val result = handler(TaskListInput)
        check(result.success && result.data != null) {
            "test: handler should return a successful task list"
        }
        check(result.data.tasks.size == 2) {
            "test: task list should return all persisted tasks"
        }
        check(result.data.tasks.first().id == "task_002") {
            "test: task list should sort by createdAt descending"
        }
    }

    override suspend fun _validate(input: TaskListInput) {
        try {
            domainCase.validate(input)
        } catch (error: IllegalArgumentException) {
            throw AppCaseError(
                code = "VALIDATION_FAILED",
                message = error.message ?: "task_list validation failed",
            )
        }
    }

    override fun _repository(): Any = resolveTaskStore()

    override suspend fun _service(input: TaskListInput): TaskListOutput {
        val taskStore = resolveTaskStore()
        val decodedTasks =
            taskStore.read().map { payload ->
                try {
                    json.decodeFromJsonElement(Task.serializer(), payload)
                } catch (error: Throwable) {
                    throw AppCaseError(
                        code = "INTERNAL",
                        message = error.message ?: "Persisted task data is invalid",
                    )
                }
            }

        assertTaskCollection(decodedTasks, "task_list.persisted_tasks")

        val output =
            TaskListOutput(
                tasks = decodedTasks.sortedByDescending { it.createdAt },
            )
        domainCase.validateOutput(output)
        return output
    }

    @Suppress("UNCHECKED_CAST")
    private fun resolveTaskStore(): TaskStore {
        val providers = ctx.extra?.get("providers") as? Map<String, Any?>
        val rawStore = providers?.get("taskStore") as? Map<String, Any?>
            ?: throw AppCaseError(
                code = "INTERNAL",
                message = "task_list requires a configured taskStore provider",
            )

        return TaskStore(
            read = rawStore["read"] as? (suspend () -> List<JsonObject>)
                ?: throw AppCaseError("INTERNAL", "task_list taskStore.read is missing"),
            write = rawStore["write"] as? (suspend (List<JsonObject>) -> Unit)
                ?: throw AppCaseError("INTERNAL", "task_list taskStore.write is missing"),
            reset = rawStore["reset"] as? (suspend () -> Unit),
        )
    }
}
